### PR TITLE
Add search callbacks to files SDK

### PR DESCRIPTION
**Pull request by awharn**
**What It Does**

Adds the ability to provide a function to the Data Sets search function, which is called if present and provided a list of data sets that are about to be searched. This allows extenders to prompt users asking if they want to continue a search, or automatically cancel a search if too many items are about to be searched. If not supplied, the search SDK keeps current behavior.

**How to Test**

**Review Checklist**
I certify that I have:
- [x] tested my changes
- [x] added/updated automated tests
- [x] updated the changelog
- [x] followed the [contribution guidelines](https://github.com/zowe/zowe-cli/blob/master/CONTRIBUTING.md)

**Additional Comments**

### CODE REVIEW

File: packages/zosfiles/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 All notable changes to the Zowe z/OS files SDK package will be documented in this file.
 
+## Recent Changes
+
+- Enhancement: Added an optional `continueSearch` function to `ISearchOptions` to allow for user prompting and data set search cancellation after the list of data sets to be searched is retrieved. [#2300](https://github.com/zowe/zowe-cli/pull/2300)
```
Comment (anaxceron) [inline packages/zosfiles/CHANGELOG.md:7]: I suggest rewording to something like: Added an optional `continueSearch` function to the `ISearchOptions` prompt/function/whatever. After an initial data set search has been executed, the new function can be set to prompt users to continue the existing search or cancel it.

**Review by anaxceron — COMMENTED** [review]
Suggested an edit to the changelog item.

File: packages/zosfiles/CHANGELOG.md (hunk 1 above)
Comment (awharn) [reply packages/zosfiles/CHANGELOG.md]: I made a change to try and follow that suggestion.

**Review by t1m0thyj — APPROVED** [review]
LGTM, thanks @awharn!

File: packages/zosfiles/src/methods/search/Search.ts
Comment (traeok) [inline packages/zosfiles/src/methods/search/Search.ts:1]: Thanks for adding comments here for clarification about the use of `cloneDeep` 😋

**Review by traeok — APPROVED** [review]
LGTM, thanks Andrew for the enhancement!

File: packages/zosfiles/src/methods/search/Search.ts [hunk 2]
```diff
@@ -100,14 +102,33 @@ export class Search {
...
             } catch (err) {
                 failedDatasets.push(pds);
             }
         }
 
+        // Call the callback if it exists.
+        let resp: boolean = true;
+        if (searchOptions.continueSearch) {
+            // Call with a clone of the search items queue. Extenders should not be modifying the original.
+            resp = await searchOptions.continueSearch(cloneDeep(searchItemsQueue));
```
Comment (gejohnston) [inline packages/zosfiles/src/methods/search/Search.ts:117]: Since the callback is someone else's code (which might crash), should we put a try-catch safety net around the callback? Certainly a stack trace will identify if the failure occurred in the callback, but the failure could be perceived as a failure in our app/API.

I know we have numerous other callbacks, and maybe they do not check for crashes, but maybe we should start? This is not a hard request for change. If you consider the idea and make a different choice, that is ok.

**Review by gejohnston — APPROVED** [review]
I made one comment for your consideration, but I still approve.

**Review by zFernand0 — COMMENTED** [review]
LGTM! 😋 

I kind of agree with Gene's comment here that we should probably catch errors coming from the callback 😋

File: packages/zosfiles/src/methods/search/Search.ts (hunk 2 above)
Comment (awharn) [reply packages/zosfiles/src/methods/search/Search.ts:117]: How would we want to handle a crash? A graceful search termination, or just re-throwing it?

Comment (awharn) [reply packages/zosfiles/src/methods/search/Search.ts:117]: Spoke with many members of the team. We decided to let the code throw any failed callback, and expect anyone using the search APIs to handle their own errors. If a callback does throw, the error should point directly to the callback function.
